pkg: detect test-only dependencies in lockdir

gridbugs opened the pull request:
Adds a new command `dune pkg list-locked-dependencies` which lists the locked dependencies of local packages, identifying which dependencies will only be included if tests are enabled.

Alizter commented:
What's the difference between this and `dune describe pkg lock`?

gridbugs commented:
> What's the difference between this and `dune describe pkg lock`?

Will update this next week to just add "test-only" annotations to the output of `dune describe pkg lock` rather than adding a new command. I forgot we had that.

Alizter commented:
It might be helpful to include the command "describe pkg lock "both in the pkg and describe command group. That way users won't need to remember which one it is in.

rgrinberg commented:
@gridbugs please undraft once this is updated

gridbugs commented:
This PR will be greatly simplified by https://github.com/ocaml/dune/pull/9172 which consolidates all the lockdir validation. I'll rebase on top of that and undraft once https://github.com/ocaml/dune/pull/9172 is merged.

gridbugs committed:
pkg: identify test-only dependencies

Adds a new command `dune pkg list-locked-dependencies` which lists the
locked dependencies of local packages, identifying which dependencies
will only be included if tests are enabled.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs commented:
I've updated this to work with the changes introduced in https://github.com/ocaml/dune/pull/9172. This is now ready for review.

gridbugs commented:
@Alizter I looked at `describe pkg lock` but that prints entire lockdirs whereas the new command in this change just prints the dependencies of individual packages according to the lockdirs. The distinction is important because a package may be a test-only dependency of one package but a non-test dependency of another.

rgrinberg commented on bin/pkg/list_locked_dependencies.ml:
```diff
@@ -0,0 +1,113 @@
+      ~traverse
+  in
+  Pp.concat
+    ~sep:Pp.newline
```
This should be `Pp.cut`. We should almost never use `Pp.newline` because it breaks that are already open.

gridbugs replied:
It looks like `Pp.cut` doesn't guarantee that the line will be broken though. My intention here is to print a nested list of titled lists of the form:
```
- Dependencies of local package a.dev
  - dep1.0.0.1
  - dep2.0.0.1
  - ...
- Dependencies of local package b.dev
  - dep1.0.0.1
  - ...
- ...
```
I can mostly achieve this with `Pp.enumerate`  but the title of the sublists (the `Dependencies of local package ...` part) isn't an element of the list so I need another way to print a line of text above the list.

What does it mean to "break boxes that are already open"?

gridbugs replied:
I've got something that I think works using `Pp.vbox` to ensure that all the `cut`s cause a break. Let me know what you think.

rgrinberg replied:
>  It looks like Pp.cut doesn't guarantee that the line will be broken though

If you want to try guarantee that a line will be broken, make a vertical box and then every space will be interpreted as a newline. See the implementation of `Pp.enumerate` for an example.

> What does it mean to "break boxes that are already open"?

Whenever you do `Pp.{box,vbox,hbox,hovbox,..} foo` we allow whoever is rendering `foo` to choose how to interpret `Pp.space` and `Pp.cut`. That allows us our errors to be well wrapped no matter the size of the user's terminal or if they're displayed as LSP diagnostics, to be wrapped there appropriately as well.

Therefore whenever you output a newline, you're breaking the layout engine's assumptions on how much more its allowed to render on this line before wrapping. Essentially, making your `Pp.t` value uncomposable since it cannot be wrapped with boxes.

rgrinberg commented on bin/pkg/list_locked_dependencies.ml:
```diff
@@ -0,0 +1,113 @@
+  in
+  Pp.concat
+    ~sep:Pp.newline
+    [ Pp.verbatim
```
How come it's not `Pp.text`?

gridbugs replied:
`Pp.text` wraps the text in a way that I find hard to read. With `verbatim` the line doesn't get broken whereas switching to `Pp.text` it wraps like:
```
- Immediate dependencies of local package
   local_2.dev
   - baz.0.0.1
   - qux.0.0.1 (test only)
```
I find that hard to read because it looks to me like the subheading is `Immediate dependencies of local package` but actually the subheading is `Immediate dependencies of local package local_2.dev`.

rgrinberg replied:
I think that's because you just forgot the box around the text. E.g. it should be:

```
Pp.hovbox (Pp.text ..)
```

We're getting `Pp.paragraph` to make this a little easier to remember.

gridbugs replied:
Ah cool I just found `Pp.hbox` which prevents line breaks.

rgrinberg commented on src/dune_pkg/package_universe.mli:
```diff
@@ -3,12 +3,26 @@ open! Import
+
+(** Raises a [User_error.E] if the lockdir doesn't contain a solution to the
+    local packages *)
+val create_exn : Local_package.t Package_name.Map.t -> Lock_dir.t -> t
```
I see this function is only used once. So I would just inline it since it's just as well implemented on top of the existing `create`.

gridbugs replied:
done

rgrinberg committed:
use Top_closure

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg commented:
I pushed a commit that uses `Top_closure`. It seems to work as we'd expect.

rgrinberg commented:
You could also consider adding this command under `dune describe pkg list-locked-dependencies`.
